pilot/pkg/xds/cds.go

```diff
@@ -92,8 +92,13 @@ func cdsNeedsPush(req *model.PushRequest, proxy *model.Proxy) bool {
 		return true
 	}
 	for config := range req.ConfigsUpdated {
-		if features.FilterGatewayClusterConfig && proxy.Type == model.Router {
-			if _, f := pushCdsGatewayConfig[config.Kind]; f {
+		if proxy.Type == model.Router {
+			if features.FilterGatewayClusterConfig {
+				if _, f := pushCdsGatewayConfig[config.Kind]; f {
+					return true
+				}
+			}
+			if proxy.MergedGateway != nil && proxy.MergedGateway.ContainsAutoPassthroughGateways && config.Kind == kind.Gateway {
```

Comment:
I am not sure I fully understand.

So the relevant types here are PeerAuth and DestinationRule. These are not in `skippedCdsConfigs`, so they are not skipped UNLESS `features.FilterGatewayClusterConfig` is true. This PR changes everything, not just `features.FilterGatewayClusterConfig`, which seems unexpected.

Additionally, it makes it so we unconditionally push on any change. Can we just add PA and DR to `pushCdsGatewayConfig` for `ContainsAutoPassthroughGateways` gateways?

```
			if features.FilterGatewayClusterConfig {
				if _, f := pushCdsGatewayConfig[config.Kind]; f {
					return true
				}
        if proxy.MergedGateway != nil && proxy.MergedGateway.ContainsAutoPassthroughGateways { // push if DR / PA changed }
			}
```

Reply:
Hm, I missed `config.Kind == kind.Gateway` actually. I don't get it, I thought the change was to PA/DR, not Gateway?

Reply:
> I am not sure I fully understand.

Ok, now I see that my description of this PR may be confusing, because it's not fully correct. Let me explain once again.

> So the relevant types here are PeerAuth and DestinationRule.

First of all, everything is good with `PeerAuthentication` and I know that it correctly triggers CDS. Actually, the logic of building clusters is problematic. Look at [buildClusters](https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/cluster.go#L252) - it returns SNI DNAT clusters only if the requesting proxy is an auto-passthrough gateway. On the other hand, creating a `Gateway` triggers only LDS and this means that after creating auto-passthrough Gateway it will not work until something else triggers CDS - for example enabling mTLS.

The above in pseudocode:

Case 1:
1. Create Gateway AUTO_PASSTHROUGH -> trigger `LDS({name: 0.0.0.0_15443, cluster_name: outbound_.8000_._.httpbin.default.svc.cluster.local})`
2. Create PeerAuthentiction STRICT -> trigger `CDS([{name: outbound|8000||httpbin.default..svc.cluster.local},{name:outbound_.8000_._.httpbin.default.svc.cluster.local}])`

Routing will work.

Case 2:
1. Create PeerAuthentiction STRICT -> trigger `CDS([{name: outbound|8000||httpbin.default..svc.cluster.local}])`
2. Create Gateway AUTO_PASSTHROUGH -> trigger `LDS({name: 0.0.0.0_15443, cluster_name: outbound_.8000_._.httpbin.default.svc.cluster.local})`

Requests will fail with 503 due to missing cluster.

Reply:
Ah so the check `proxy.Type == model.Router && proxy.MergedGateway != nil && proxy.MergedGateway.ContainsAutoPassthroughGateways` for `buildOutboundSniDnatClusters` is logically changing, but we don't trigger the push. That makes sense. Thanks

Reply:
It would be nice to not push CDS on any gateway change for this case if we can. Its fairly expensive.

For sidecars we have `PrevSidecarScope` which we use in `proxyDependentOnConfig` which helps for similar checks, but we don't have a `PrevMergedGateway` to do a similar check (and I am not sure we should).

I guess its not _any_ gateway since its just AUTO_PASSTHROUGH ones, but still -- now we will push to them excessively.

@ramaraochavali @hzxuzhonghu any thoughts? I don't _love_ this approach but cannot come up with a better approach, so maybe its fine. I have a general desire to improve the "needs push" stuff, but that is way out of scope for this issue.

Reply:
> It would be nice to not push CDS on any gateway change for this case if we can. Its fairly expensive.

Yes, I am aware of that, but I think that most changes applied to auto-passthrough gateways require pushing CDS, because `AUTO_PASSTHROUGH` mode is used for east-west traffic, so possibly the only frequent changes are adding hostnames to `spec.servers[].hosts` what causes 503 as well. Pushing CDS may have negative impact if a single gateway is used for ingress and east-west traffic, but I guess that most users deploy these gateway separately.

Reply:
Maybe we can fix this soon, and improve xxPush later. IMO， this may be also good to support delta cds for gateway

Reply:
Of course I can implement `PrevMergedGateway`, but could we merge this fix ASAP to include it in 1.22? And I would implement optimization in a follow-up PR?

Reply:
> Yes, I am aware of that, but I think that most changes applied to auto-passthrough gateways require pushing CDS, because AUTO_PASSTHROUGH mode is used for east-west traffic, so possibly the only frequent changes are adding hostnames to spec.servers[].hosts what causes 503 as well. Pushing CDS may have negative impact if a single gateway is used for ingress and east-west traffic, but I guess that most users deploy these gateway separately.

The impact is actually for all users of E-W. Because we will push ANY Gateway change, even if it was targetting a different (ingress) gateway. I think.

Anyhow, still probably best to be correct before we worry about optimizations